src/treelearner/cuda_tree_learner.cpp

```diff
@@ -0,0 +1,973 @@
+  size_t  copied_feature = 0;
+  // set device info
+  int device_id = 0;
+  uint8_t* device_features = device_features_[device_id];
+  CUDASUCCESS_OR_FATAL(cudaSetDevice(device_id));
+  Log::Debug("Started copying dense features from CPU to GPU - 1");
+
+  for (int i = 0; i < num_feature_groups_; ++i) {
+    // looking for dword_features_ non-sparse feature-groups
+    if (!train_data_->IsMultiGroup(i)) {
+      dense_feature_group_map_.push_back(i);
+      auto sizes_in_byte = train_data_->FeatureGroupSizesInByte(i);
+      void* tmp_data = train_data_->FeatureGroupData(i);
+           Log::Debug("Started copying dense features from CPU to GPU - 2");
+      CUDASUCCESS_OR_FATAL(cudaMemcpyAsync(&device_features[copied_feature * num_data_], tmp_data, sizes_in_byte, cudaMemcpyHostToDevice, stream_[device_id]));
+           Log::Debug("Started copying dense features from CPU to GPU - 3");
+      copied_feature++;
+      // reset device info
+      if (copied_feature == (size_t) num_gpu_feature_groups_[device_id]) {
+         CUDASUCCESS_OR_FATAL(cudaEventRecord(features_future_[device_id], stream_[device_id]));
+         device_id += 1;
+         copied_feature = 0;
+         if (device_id < num_gpu_) {
+           device_features = device_features_[device_id];
+           CUDASUCCESS_OR_FATAL(cudaSetDevice(device_id));
+         }
+      }
+    } else {
+      sparse_feature_group_map_.push_back(i);
+    }
+  }
+}
+
+
+
+// LGBM_CUDA: InitGPU w/ num_gpu
+void CUDATreeLearner::InitGPU() {
+  // Get the max bin size, used for selecting best GPU kernel
+  max_num_bin_ = 0;
+
+  #if GPU_DEBUG >= 1
+  printf("bin_size: ");
+  #endif
+  for (int i = 0; i < num_feature_groups_; ++i) {
+    if (train_data_->IsMultiGroup(i)) {
+      continue;
+    }
+    #if GPU_DEBUG >= 1
+    printf("%d, ", train_data_->FeatureGroupNumBin(i));
+    #endif
+    max_num_bin_ = std::max(max_num_bin_, train_data_->FeatureGroupNumBin(i));
+  }
+  #if GPU_DEBUG >= 1
+  printf("\n");
+  #endif
+
+  if (max_num_bin_ <= 16) {
+    device_bin_size_ = 16;
+    histogram_size_ = 16;
+    dword_features_ = 1;
+  } else if (max_num_bin_ <= 64) {
+    device_bin_size_ = 64;
+    histogram_size_ = 64;
+    dword_features_ = 1;
+  } else if (max_num_bin_ <= 256) {
+    Log::Debug("device_bin_size_ = 256");
+    device_bin_size_ = 256;
+    histogram_size_ = 256;
+    dword_features_ = 1;
+  } else {
+    Log::Fatal("bin size %d cannot run on GPU", max_num_bin_);
+  }
+  if (max_num_bin_ == 65) {
+    Log::Warning("Setting max_bin to 63 is sugguested for best performance");
+  }
+  if (max_num_bin_ == 17) {
+    Log::Warning("Setting max_bin to 15 is sugguested for best performance");
+  }
+
+  // LGBM_CUDA: get num_dense_feature_groups_
+  CountDenseFeatureGroups();
+
+  // LGBM_CUDA: initialize GPU
+  CUDASUCCESS_OR_FATAL(cudaGetDeviceCount(&num_gpu_));
+  if (num_gpu_ > 1) num_gpu_ = 1;
+  if (num_gpu_ > num_dense_feature_groups_) num_gpu_ = num_dense_feature_groups_;
```

Comment:
Rather not.  This is either actively being looked at or soon will be.  Would caused a mess merging.

Reply:
> This is either actively being looked at or soon will be.

Could you please be more specific here? What does it mean? Are you sure that the case when you find multi-GPU (for example) is inefficient or impossible to implement will be never happen?

> Would caused a mess merging.

Sorry, didn't get it. It should just significantly reduce the code for review by removing all those `for` loops through GPUs.

I'm not sure that we need unused (dead) code in the repo given that we are lacking of maintainers especially those who are familiar with GPU computing.

Reply:
Again, there is another group or people that are investigating multiple GPU.  It doesn't make sense to rip this out only to have it put back in (and it also causes major merge headaches)

Reply:
> [multi-]GPU computing.

Catboost allows you to [select a range of cards for training], (https://catboost.ai/docs/features/training-on-gpu.html) for training, but from my limited tests with this algo (it is less accurate than the other two) using more than 1 card never had any positive impact on model training times (and the company did splurge on the NVLink sockets to improve inter-GPU bandwidth and allow for better scaling). So multiple cards research in case of boosted trees is likely a dead-end if Yandex team failed to achieve any multi-GPU scaling (and they do know how their CUDA - never seen better implementation in terms of efficiency, control, and resource management, easily beats most if not all DNNs).

Reply:
> Again, there is another group or people that are investigating multiple GPUs. It doesn't make sense to rip this out only to have it put back in (and it also causes major merge headaches)

I'm afraid that `to have it put back in` will be never happen.

> (and it also causes major merge headaches)

Sorry, I can't get it. This is new file in the repo. It is impossible to have merge conflicts. Or are you speaking about your internal development cycle?

Reply:
I agree with @StrikerRUS . This project is maintained by a small group of volunteers, and we should not take on significant code that is unused and part of a "to be done at some time in the future" feature.

As long as we have the commitment to maintaining everything on `master`, we have to be protective of it.

Reply:
We disagree and feel this is valuable code.   Multi-GPUs will have benefits in certain platforms (in the future) - non-x64 or on the cloud.

Reply:
If the code will not be reachable on the `master` branch after this PR is merged, it isn't providing any value.

I don't disagree that multi-GPU training is valuable and important, but this code we're talking about doesn't provide that. It lays the foundation for that but more work would have to be done to make it usable, right?

Reply:
It code for using multiple GPUs on CUDA works and is reachable.  It allows for training on more than one GPU if the condition preventing more than one GPUs is eliminated.

Reply:
@ChipKerchner

> We disagree and feel this is valuable code. Multi-GPUs will have benefits in certain platforms (in the future) - non-x64 or on the cloud.

XGBoost has dropped support of single node multi-GPU support recently and switched to Dask as an underlying technology: https://github.com/dmlc/xgboost/issues/4531. And given the comment above from @mirekphd (https://github.com/microsoft/LightGBM/pull/3160#discussion_r476686213), this is quite arguable statement.

> The code for using multiple GPUs on CUDA works and is reachable. It allows for training on more than one GPU if the condition preventing more than one GPUs is eliminated.

OK. But why then this restriction is in the code now? What is preventing to eliminate it right now?